Clear AR connections in tests before forking for parallelization

flavorjones opened the pull request:
### Motivation / Background

Works around #41176 which seems to affect the mysql2 adapter built with the mysql8 client library on MacOS.

### Detail

This PR introduces a before-fork hook into ActiveSupport::Testing::Parallelization, which is invoked immediately before spawning the worker processes.

This PR also implements a before-fork hook in ActiveRecord::TestDatabases to clear connections, to avoid the symptoms of poor fork-safety in the mysql2 8.0 client / mysql2 adapter.

We see this problem relatively often at 37s when running large test suites on MacOS development machines.

### Additional information

This PR does not directly address the fork-safety issue in the mysql8 client library, but closing connections in a test process before forking doesn't have any obvoius negative impacts that I know of, and so my hope is something like this would be an acceptable tradeoff.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [-] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

flavorjones committed:
Clear AR connections in tests before forking for parallelization

Fixes #41176 which seems to affect the mysql2 adapter on MacOS.

The underlying mysql problem was diagnosed in this comment by
@jgloudon:

https://github.com/rails/rails/issues/41176#issuecomment-1492649116

It seems like on MacOS the mysql8 client library is using kqueue which
creates a file descriptor that is not carried into the child process,
allowing for the FD to be re-used.

Co-authored-by: Donal McBreen <donal@basecamp.com>

flavorjones commented:
An interesting historical note: I tried previously to introduce some pre-fork behavior to Rails generally in https://github.com/rails/rails/pull/52931#issuecomment-2351365601 to address an unrelated fork-safety issue with sqlite.

The thoughts at that time were that we don't want to mess with parent processes' open connections because it requires stopping the world (i.e., getting a lock on all connections might be hard to do in a timely fashion on a live web server). Which I think is a reasonable worry and I was able to create an alternative post-fork fix that lives inside the sqlite3-ruby gem.

cc @dhh who has thoughts about whether/when/if/how we should allow the framework to close connections before forking.

byroot commented on activesupport/lib/active_support/testing/parallelization.rb:
```diff
@@ -32,7 +40,14 @@ def initialize(worker_count)
       end
 
+      def before_fork
+        Parallelization.before_fork_hooks.each do |cb|
+          cb.call
+        end
```
```suggestion
        Parallelization.before_fork_hooks.each(&cb)
```

/nitpick

byroot commented on activerecord/lib/active_record/test_databases.rb:
```diff
@@ -4,6 +4,11 @@
 module ActiveRecord
   module TestDatabases # :nodoc:
+    ActiveSupport::Testing::Parallelization.before_fork_hook do
+      puts "PREFORK: clearing connections"
```
This output may be annoying.

flavorjones replied:
ugh, this was unintentional! 🤦

zzak replied:
It's a feature.

byroot reviewed (commented):
I don't have any strong opinion against it.

Regardless of your issues with mysql2, a pre-fork hook is always sensible as users may need to do other fork preparation.

A post fork hook would probably make sense too.

That being said, I would have prefered to figure out what the `mysql2` problem actually is, but given how shitty `libmysqlclient` is, and that we know have `trilogy`, it might not be worth the effort. If anything I'd rather start considering a timeline to extract `Mysql2Adapter` into its own gem, and winding down support.

byroot commented:
Just to give some more actionable feedback. I think we shouldn't just provide `before_fork`, but also `after_fork` for consistency with most other fork callback systems. Other than that I'm OK with this change.

matthewd commented:
Looks like `after_fork_hook` is already there.

So yeah I think ✂️ the `puts` and this is fine.

I'd like to imagine we could fix the underlying issue in mysql2, because this _shouldn't_ be necessary, but given it's only happening at the start of test parallelisation (where I imagine it'd be moderately rare to even have a connection to clear?)... yeah, it's fine. 😅

flavorjones commented:
> I would have prefered to figure out what the mysql2 problem actually is

The underlying issue was diagnosed in [this comment](https://github.com/rails/rails/issues/41176#issuecomment-1492649116) by @jgloudon. On MacOS the mysql8 client library is using kqueue which creates a file descriptor that is not carried into the child process, allowing for the FD to be re-used, and hilarity ensues. It's not clear if it can be worked around / fixed in the ruby driver, but that library has not had high PR velocity lately.

I've addressed the concerns raised in the review notes. Thanks for your time!
